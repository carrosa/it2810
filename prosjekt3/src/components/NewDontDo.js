import React from 'react';
import {View, StyleSheet, TouchableWithoutFeedback, AsyncStorage} from 'react-native';
// import {Button, TextInput} from 'react-native-paper';
import {Button, Input} from 'react-native-elements';


const styles = StyleSheet.create({
  // Styles for denne classen
  submitButton: {
    alignSelf: 'center',
    width: '50%',
    borderRadius: 50,
    margin: 20
  },
  view: {
    height: '100%'
  }
});

// Brukes ved onPress til å dismisse keyboard
const dismissKeyboard = require('dismissKeyboard');

export default class NewDontDo extends React.Component {

  state = {
    title: '',
    content: '',
    tasks: {}
  };

  async componentDidMount() {
    // Hvis det har blitt lagt til en dont do tidligere så henter vi IDen dens
    try {
      const tasks = await AsyncStorage.getItem('tasks');
      if (tasks) { // Sjekker om det finnes noen tasks, hvis ja så setter vi task state
        this.setState({tasks: JSON.parse(tasks)});
      }
    } catch (e) {
      console.error(e);
    }
  }

  render() {
    return (
        <TouchableWithoutFeedback onPress={dismissKeyboard}>
          {/* When clicking the screen (not  */}
          <View style={styles.view}>
            <Input
                placeholder='Tittel'
                label='Tittel på Dont Do'
                value={this.state.title}
                onChangeText={title => this.setState({title: title})}
            />
            <Input
                placeholder='Beskrivelse'
                label='Beskrivelse på Dont Do'
                value={this.state.content}
                onChangeText={content => this.setState({content: content})}
            />
            <Button
                disabled={!(this.state.title.length > 0 && this.state.content.length > 0)}
                style={styles.submitButton}
                title={'Lagre'}
                onPress={async () => {
                  await this.saveNewDontDo().then(() => {
                    this.navigateToDontDosAndUpdate();
                  });
                }}/>
          </View>
        </TouchableWithoutFeedback>
    );
  }

  constructNewDontDo() {
    // lager en ny task state og setter state til det, og asyncstorage setitem tasks til det
    const newTaskNumber = Object.keys(this.state.tasks).length;
    let newTaskState = this.state.tasks;
    /*
    * Sjekker om tittel og content er lengde på mer enn 0
    * Dette trengs egentlig ikke i appen, for knappen funker ikke hvis det er sånn
    * men pga testing blir det lettest å gjøre det på denne måten
    */
    if (this.state.title.length > 0 && this.state.content.length > 0) {
      newTaskState[`${newTaskNumber}`] = {
        'title': this.state.title,
        'content': this.state.content,
        'done': false
      };
    }
    return newTaskState;
  }

  async saveNewDontDo() {
    /*
     * Bruker AsyncStorage til å legge til en ny Dont Do
     */
    try {
      const newDontDo = this.constructNewDontDo();
      this.setState({tasks: newDontDo});
      await AsyncStorage.setItem('tasks', JSON.stringify({
        ...newDontDo
      }));

    } catch (e) {
      console.error(e);
    }
  }

  navigateToDontDosAndUpdate() {
    /*
     * Navigerer tilbake til Dont Dos
     * Popper denne instansen av NewDontDo for at appen ikke skal huske hva som du har skrevet her etter at det er
     * lagra (For å slippe å viske bort neste gang du skal legge til en ny to dont)
     * Må ha alt dette for å fjerne alt fra navigation stacken og pushe på en ny DontDos for at
     * vi ikke skal få flere DontDos som legger seg "oppå" hverandre, samtidig som vi vil at
     * DontDos sin componentDidMount metode skal kalles
     */
    this.props.navigation.dismiss();
    this.props.navigation.popToTop();
    this.props.navigation.push('DontDos');
  }
}
