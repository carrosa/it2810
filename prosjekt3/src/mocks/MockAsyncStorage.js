/*
 * Klasse for å mocke AsyncStorage under testing
 * Alle funksjonene returnerer et promise sånn at vi kan teste async
 */
export default class MockAsyncStorage {
  constructor(cache = {}) {
    this.storageCache = cache;
  }

  setItem = jest.fn((key, value) => {
    return new Promise((resolve, reject) => {
      return (typeof key !== 'string' || typeof value !== 'string')
          ? reject(new Error('key and value must be string'))
          : resolve(this.storageCache[key] = value);
    });
  });

  getItem = jest.fn((key) => {
    return new Promise((resolve) => {
      return this.storageCache.hasOwnProperty(key)
          ? resolve(this.storageCache[key])
          : resolve(null);
    });
  });
}