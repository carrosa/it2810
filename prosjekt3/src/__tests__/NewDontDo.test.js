import React from 'react';
import renderer from 'react-test-renderer';
import NewDontDo from '../components/NewDontDo';
import MockAsyncStorage from '../mocks/MockAsyncStorage';

/*
* Mock states vi tester med (setter state til instance med disse)
*/
const noPreviousTasksTestState = {
  title: 'Test Title',
  content: 'Test Content',
  tasks: {}
};
const multiplePreviousTasksTestState = {
  title: 'Test Title',
  content: 'Test Content',
  tasks: {
    0: {
      'title': 'previous title 1',
      'content': 'previous content 1',
      'done': false
    },
    1: {
      'title': 'previous title 1',
      'content': 'previous content 1',
      'done': false
    }
  }
};

/*
* Expected return for testen til saveNewDontDo
*/
const expectedAsyncGetTasks = {

};

/*
 * Setter opp AsyncStorage Mock
 */
const storageCache = {};
const AsyncStorage = new MockAsyncStorage(storageCache);
jest.setMock('AsyncStorage', AsyncStorage);



describe('<NewDontDo /> tests', () => {
  // Bygger NewDontDo
  const wrapper = renderer.create(<NewDontDo/>);
  const instance = wrapper.getInstance();

  // Test bygge et snapshot uten noe data
  test('<NewDontDo /> renders correctly', () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  /*
   * Test Lage et nytt dont do object når vi ikke har noen fra før
   * Setter først statene vi skal teste for. Både med ingen previous tasks og med flere previous tasks
   */
  // Test med ingen previous tasks
  test('constructNewDontDo noPreviousState', () => {
    instance.state = noPreviousTasksTestState;
    expect(instance.constructNewDontDo()).toMatchSnapshot();
  });
  // Test med flere previous tasks
  test('constructNewDontDo multiplePreviousState', () => {
    instance.state = multiplePreviousTasksTestState;
    expect(instance.constructNewDontDo()).toMatchSnapshot();
  });

  /*
   * Test saveNewDontDo with mocked AsyncStorage
   * saveNewDontDo blir kalt med state satt i forrige test
   * Altså med state: multiplePreviousTasksTestState
   *
   * Forvented return:
   * stringified JSON av det som constructNewDontDo() spytter ut
   * den metoden er testet i forrige test, så vi vet at den funker
   *
   * Først: Test at tasks er tom
   */
  test('AsyncStorage tasks is empty', async () => {
    expect(await AsyncStorage.getItem('tasks')).toBe(null);
  });
  test('saveNewDontDo tests', async () => {
    await instance.saveNewDontDo();
    const tasks = await AsyncStorage.getItem('tasks');
    expect(tasks).toMatch(JSON.stringify(instance.constructNewDontDo()))
  });
});
