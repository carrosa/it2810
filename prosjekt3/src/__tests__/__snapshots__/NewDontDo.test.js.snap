// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewDontDo /> tests <NewDontDo /> renders correctly 1`] = `
<View
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "height": "100%",
    }
  }
>
  <View
    style={
      Object {
        "width": "90%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#86939e",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Tittel på Dont Do
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Tittel"
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "flex": 1,
            "fontSize": 18,
            "height": 40,
            "marginLeft": 10,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    style={
      Object {
        "width": "90%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#86939e",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Beskrivelse på Dont Do
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 1,
          "borderColor": "#86939e",
          "flexDirection": "row",
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Beskrivelse"
        style={
          Object {
            "alignSelf": "center",
            "color": "black",
            "flex": 1,
            "fontSize": 18,
            "height": 40,
            "marginLeft": 10,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    style={
      Array [
        undefined,
        false,
      ]
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "center",
          "borderRadius": 50,
          "margin": 20,
          "opacity": 1,
          "width": "50%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#D1D5D8",
            "borderRadius": 3,
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "backgroundColor": "transparent",
              "color": "#F3F4F5",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            }
          }
        >
          Lagre
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<NewDontDo /> tests constructNewDontDo multiplePreviousState 1`] = `
Object {
  "0": Object {
    "content": "previous content 1",
    "done": false,
    "title": "previous title 1",
  },
  "1": Object {
    "content": "previous content 1",
    "done": false,
    "title": "previous title 1",
  },
  "2": Object {
    "content": "Test Content",
    "done": false,
    "title": "Test Title",
  },
}
`;

exports[`<NewDontDo /> tests constructNewDontDo noPreviousState 1`] = `
Object {
  "0": Object {
    "content": "Test Content",
    "done": false,
    "title": "Test Title",
  },
}
`;
